velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,231 @@
+    for (vector_size_t i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      clearNull(group);
+      rawValues[i] = StringView(value<char>(group), sizeof(int64_t));
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    VELOX_CHECK_EQ((*result)->encoding(), VectorEncoding::Simple::FLAT);
+    auto vector = (*result)->as<FlatVector<int64_t>>();
+    VELOX_CHECK(
+        vector,
+        "Unexpected type of the result vector: {}",
+        (*result)->type()->toString());
+    vector->resize(numGroups);
+
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      if (isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        vector->set(i, *value<int64_t>(group));
+      }
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    BufferPtr hashes =
+        AlignedBuffer::allocate<int64_t>(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.template applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
+        computeHashForNull(groups[row]);
+      } else {
+        computeHash(groups[row], rawHashes, row);
+      }
+    });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    VELOX_CHECK_EQ(args[0]->encoding(), VectorEncoding::Simple::FLAT);
+
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected(
+        [&](vector_size_t i) { *value<int64_t>(groups[i]) += rawValues[i]; });
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    BufferPtr hashes =
+        AlignedBuffer::allocate<int64_t>(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.template applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
+        computeHashForNull(group);
+      } else {
+        computeHash(group, rawHashes, row);
+      }
+    });
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    VELOX_CHECK_EQ(args[0]->encoding(), VectorEncoding::Simple::FLAT);
+    clearNull(group);
+
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+    int64_t result = 0;
+    if (vector->mayHaveNulls()) {
```

Comment:
I'm just curious about the difference between `addIntermediateResults()` and `addSingleGroupIntermediateResults()`. I thought the two functions were similar except the former deals with multiple groups whereas the latter deals with a single group. So I thought they either both check for null or both do not. Are they actually more different?

Reply:
Thats a good question/observation Wei, I have noticed that singe group intermediate results can have nulls in them : for e.g if you have multiple splits, some filter condition , then no aggregation takes place in some of the splits, there single group intermediate results is called once / split (and can be called potentially on a different machine). However I havent noticed this for addIntermediateResults. countAggregate also follows the same pattern. Maybe I should add null checks in addIntermediateResults too.